pkg/nsxt/nsxt_connector.go

```diff
@@ -0,0 +1,155 @@
+/*
+ Copyright 2020 The Kubernetes Authors.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package nsxt
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/vmware/vsphere-automation-sdk-go/runtime/core"
+	"github.com/vmware/vsphere-automation-sdk-go/runtime/protocol/client"
+	"github.com/vmware/vsphere-automation-sdk-go/runtime/security"
+	"k8s.io/cloud-provider-vsphere/pkg/nsxt/config"
+)
+
+// NewNsxtConnector creates a new NSXT connector
+func NewNsxtConnector(nsxtConfig *config.NsxtConfig) (client.Connector, error) {
+	url := fmt.Sprintf("https://%s", nsxtConfig.Host)
+	securityCtx := core.NewSecurityContextImpl()
+	securityContextNeeded := true
+	if len(nsxtConfig.ClientAuthCertFile) > 0 {
+		securityContextNeeded = false
+	}
+
+	if securityContextNeeded {
+		if len(nsxtConfig.VMCAccessToken) > 0 {
+			if nsxtConfig.VMCAuthHost == "" {
+				return nil, fmt.Errorf("vmc auth host must be provided if auth token is provided")
+			}
+
+			apiToken, err := getAPIToken(nsxtConfig.VMCAuthHost, nsxtConfig.VMCAccessToken)
+			if err != nil {
+				return nil, err
+			}
+
+			securityCtx.SetProperty(security.AUTHENTICATION_SCHEME_ID, security.OAUTH_SCHEME_ID)
+			securityCtx.SetProperty(security.ACCESS_TOKEN, apiToken)
+		} else {
+			if nsxtConfig.User == "" {
+				return nil, fmt.Errorf("username must be provided")
+			}
+
+			if nsxtConfig.Password == "" {
+				return nil, fmt.Errorf("password must be provided")
+			}
+
+			securityCtx.SetProperty(security.AUTHENTICATION_SCHEME_ID, security.USER_PASSWORD_SCHEME_ID)
+			securityCtx.SetProperty(security.USER_KEY, nsxtConfig.User)
+			securityCtx.SetProperty(security.PASSWORD_KEY, nsxtConfig.Password)
+		}
+	}
+
+	tlsConfig, err := getConnectorTLSConfig(nsxtConfig.InsecureFlag, nsxtConfig.ClientAuthCertFile, nsxtConfig.ClientAuthKeyFile, nsxtConfig.CAFile)
+	if err != nil {
+		return nil, err
+	}
+	httpClient := http.Client{
+		Transport: &http.Transport{
+			Proxy:           http.ProxyFromEnvironment,
+			TLSClientConfig: tlsConfig,
+		},
+	}
+
+	connector := client.NewRestConnector(url, httpClient)
+	connector.SetSecurityContext(securityCtx)
+	return connector, nil
+}
+
+func getConnectorTLSConfig(insecure bool, clientCertFile string, clientKeyFile string, caFile string) (*tls.Config, error) {
```

Comment:
I think it would be good to add some comments here, same for `getAPIToken` function